### PR TITLE
chore(provider): SailPoint Identity Secure Cloud (ISC)

### PR DESCRIPTION
**Pull request by mike818148**
## ☕️ Reasoning

A new oauth provider for SailPoint Identity Secure Cloud (ISC). Developers then can utilize SailPoint Identity Secure Cloud (ISC) to authenticate the custom authentication based on OAuth2 standard. 

## 🧢 Checklist

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)
- [SailPoint Identity Secure Cloud - Authentication](https://developer.sailpoint.com/docs/api/authentication)
- [SailPoint Identity Secure Cloud - Authorization](https://developer.sailpoint.com/docs/api/authorization)
- [SailPoint Developer Community](https://developer.sailpoint.com/discuss/)
- [Learn more about OAuth](https://authjs.dev/concepts/oauth)

### CODE REVIEW

**Comment by mike818148** [comment]
Hello, 

please let me know should I also prepare a test environment for your to conduct some testing, understand it will be hard to allocate a SailPoint IdentityNow tenant for you to test directly. By providing this provider, I believe it will bring some value to the SailPoint Developer community, especially for React developers.

**Comment by ndom91** [comment]
@mike818148 thanks for the interest in the project. 

So if I understand correctly, SailPoint is an enterprise identity/security platform? So a random developer like me wouldn't be able to get OAuth Client credentials probably, right?

**Comment by ndom91** [comment]
In the mean time, you could also add a docs page at `/docs/pages/getting-started/providers/sailpoint.mdx`. You can copy the structure of any of the other providers, like `Netlify.mdx` for example, and just fill in the info yuo've got the JSDoc of your provider already :pray:

**Comment by mike818148** [comment]
> @mike818148 thanks for the interest in the project.
> 
> So if I understand correctly, SailPoint is an enterprise identity/security platform? So a random developer like me wouldn't be able to get OAuth Client credentials probably, right?

Hello @ndom91, 

regretfully, this is correct, SailPoint Identity Secure Cloud(ISC) is an enterprise SaaS platform for identity/security solution. At the moment, random developer may not get an OAuth Client credentials (from a certain tenant) as easy as other OAuth provider, however, I believe this statement is also changing. As per my understanding, SailPoint already have a [open developer community](https://developer.sailpoint.com/discuss/) which welcome public to join, and it also has a [opensource project ](https://github.com/sailpoint-oss), and as an active developer community member, there is a way to requst a ISC tenant for demo/developing.

**Comment by ndom91** [comment]
> Hello @ndom91, 
> 
> this is correct, SailPoint Identity Secure Cloud(ISC) is an enterprise SaaS platform for identity/security solution. At the moment, random developer may not get an OAuth Client credentials (from a certain tenant) as easy as other OAuth provider, however, I believe this statement is also changing. As per my understanding, SailPoint already have a [open developer community](https://developer.sailpoint.com/discuss/) which welcome public to join, and it also has a [opensource project ](https://github.com/sailpoint-oss), and as an active developer community member, there is a way to requst a ISC tenant for demo/developing. 

Oh okay I see. If you could request a demo tenant for this so we can continuously test it etc. that'd be amazing 🙏

**Comment by mike818148** [comment]
> > Hello @ndom91,
> > this is correct, SailPoint Identity Secure Cloud(ISC) is an enterprise SaaS platform for identity/security solution. At the moment, random developer may not get an OAuth Client credentials (from a certain tenant) as easy as other OAuth provider, however, I believe this statement is also changing. As per my understanding, SailPoint already have a [open developer community](https://developer.sailpoint.com/discuss/) which welcome public to join, and it also has a [opensource project ](https://github.com/sailpoint-oss), and as an active developer community member, there is a way to requst a ISC tenant for demo/developing.
> 
> Oh okay I see. If you could request a demo tenant for this so we can continuously test it etc. that'd be amazing 🙏
> 
> If it goes through, you can DM it to me on our [discord](https://discord.authjs.dev) for example (`@ndom91`)

thanks a lot. I already have a tenant, would it be okai, we use this tenant for testing, if yes, I just need your email to register on the tenant. This tenant only contains mock-up data. I will then also share the guide regarding OAuth client setup with you (just to show you where to find it)

**Comment by mike818148** [comment]
> mdx

Hello @ndom91, 

just add the file sailpoint.mdx, question here: for the environement variables not sure does this AUTH_[Provider]_* will also work in my case, may you express more on this, how this is working? Otherwise, I will just document to pass the value directly. 

To be honested, I only test this code with NextJS, not yet on other frameworks (Svelte and Express) 

In addition, I am preparing a blog for the use cases, once published, will update the sailpoint.mdx with corresponding link. 

Thanks and Regards,

File: docs/pages/getting-started/providers/sailpoint.mdx [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+import { SvelteKitAuth } from "@auth/sveltekit"
+import SailPoint from "@auth/sveltekit/providers/sailpoint"
+
+export const { handle, signIn, signOut } = SvelteKitAuth({
+  providers: [
+    {
+      id: "identitySecureCloud",
+      name: "Identity Secure Cloud",
+      type: "oauth",
+      clientId: process.env.ISC_CLIENT_ID!,
```
Comment (ndom91) [inline docs/pages/getting-started/providers/sailpoint.mdx:106]: The sveltekit environment variables work a bit different, check out this example:

image

Comment (mike818148) [reply docs/pages/getting-started/providers/sailpoint.mdx]: Hello, 

Thanks a lot for the correction. Just made the change in the PR.

File: docs/pages/getting-started/providers/sailpoint.mdx [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+```bash
+https://example.com/api/auth/callback/identitySecureCloud
+```
+
+  </Code.Next>
+  <Code.Svelte>
+
+```bash
+https://example.com/api/auth/callback/identitySecureCloud
```
Comment (ndom91) [inline docs/pages/getting-started/providers/sailpoint.mdx:31]: In all other packages the default dynamic route is at `/auth/*` not `/api/auth/*`, small difference.

```suggestion
https://example.com/auth/callback/identitySecureCloud
```

File: docs/pages/getting-started/providers/sailpoint.mdx [hunk 3]
```diff
@@ -0,0 +1,255 @@
+import { Callout } from "nextra/components"
+import { Code } from "@/components/Code"
+
+<img align="right" src="/img/providers/sailpoint.svg" height="64" width="64" />
+
+# SailPoint Identity Secure Cloud Provider
+
+SailPoint Identity Secure Cloud (ISC) is an enterprise SaaS platform for identity/security solution. In order to test with this OAuth integration, you will need an ISC tenant, if your are a sailpoint customer or partner, you may inquiry with SailPoint partner/customer success manager. If you are a developer, you may inquiry more information on [SailPoint Developer Community](https://developer.sailpoint.com/discuss/) for more information.
```
Comment (ndom91) [inline docs/pages/getting-started/providers/sailpoint.mdx:8]:
```suggestion
SailPoint Identity Secure Cloud (ISC) is an enterprise SaaS platform for identity and security. In order to use this OAuth integration, you will need an ISC tenant. If you're a SailPoint customer or partner, please talk to your SailPoint account manager for more details. If you are a developer, you can check out the [SailPoint Developer Community](https://developer.sailpoint.com/discuss/). 
```

File: docs/pages/getting-started/providers/sailpoint.mdx [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+```bash
+https://example.com/auth/callback/identitySecureCloud
+```
+
+  </Code.Svelte>
+</Code>
+
+### Create OAuth Client
+
+Find your Identity Secure Cloud Tenant OAuth Information (can be retrieved under https://{tenant}.api.identitynow.com/oauth/info). Create OAuth Client ([guide](https://documentation.sailpoint.com/saas/help/common/api_keys.html?h=oauth+client#creating-an-api-key))with grant types: AUTHORIZATION_TOKEN and REFRESH_TOKEN, Redirect URL: Callback URL above,   select scopes: sp:scope:all. Note down the clientId and clientSecret.
```
Comment (ndom91) [inline docs/pages/getting-started/providers/sailpoint.mdx:39]:
```suggestion
Find your Identity Secure Cloud Tenant OAuth Information which can be found at `https://{tenant}.api.identitynow.com/oauth/info`. Create an OAuth Client (following this [guide](https://documentation.sailpoint.com/saas/help/common/api_keys.html?h=oauth+client#creating-an-api-key)) with grant types: `AUTHORIZATION_TOKEN` and `REFRESH_TOKEN`. Redirect URL should match your version of the Callback URL above. Finally, select the scopes `sp:scope:all`. Note down the generated `clientId` and `clientSecret`.
```

File: docs/pages/getting-started/providers/sailpoint.mdx [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+      },
+      style: { text: "#011E69", bg: "#fff", logo: "sailpoint.svg" },
+    },
+] }))
+```
+
+  </Code.Express>
+</Code>
+
+### Configuration for Refresh Token
```
Comment (ndom91) [inline docs/pages/getting-started/providers/sailpoint.mdx:177]: Is there anything about this refresh process that's different with SailPoint? If not, lets drop this section or just link to the existing guide. If there are significant differences, then lets point those out.

Comment (mike818148) [reply docs/pages/getting-started/providers/sailpoint.mdx:177]: I think we can drop this section off first, gernally it's the same as the others, was just thinking provider some example there.

Comment (mike818148) [reply docs/pages/getting-started/providers/sailpoint.mdx]: Just removed it. Thanks.

File: docs/pages/getting-started/providers/sailpoint.mdx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+      type: "oauth",
+      clientId: process.env.ISC_CLIENT_ID!,
+      clientSecret: process.env.ISC_CLIENT_SECRET!,
+      authorization: {
+        url: `${process.env.ISC_BASE_URL!}/oauth/authorize`,
+        params: { scope: 'sp:scopes:all' },
+      },
+      token: `${process.env.ISC_BASE_API_URL!}/oauth/token`,
+      userinfo: `${process.env.ISC_BASE_API_URL!}/oauth/userinfo`,
+      profile(profile: IdentitySecureCloudProfile) {
```
Comment (ndom91) [inline docs/pages/getting-started/providers/sailpoint.mdx:154]: This should be somethign like this, for all 3 examples

```ts
      profile(profile: IdentitySecureCloudProfile) {
        return {
            id: profile.id,
            email: profile.email,
            name: `${profile.firstname} ${profile.lastname}`,
            image: null
        }
      },
```

File: docs/pages/getting-started/providers/sailpoint.mdx [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            name: profile.uid,
+        }
+      },
+      style: { text: "#011E69", bg: "#fff", logo: "sailpoint.svg" },
+    },
+] }))
+```
+
+  </Code.Express>
+</Code>
```
Comment (ndom91) [inline docs/pages/getting-started/providers/sailpoint.mdx:175]: And somethign like this below the example code to point folks in the right direction, should they want to use any of the additional fields

```suggestion
</Code>

Your `userprofile` endpoint will return more fields, but by default the [User table](https://authjs.dev/getting-started/database#models) only supports `id`, `name`, `email`, and `image`. Therefore, if you'd like to use any of the following fields, make sure you modify the `User` table schema in whichever adapter / database you're using.

	```ts
	tenant: profile.tenant,
	id: profile.id,
	uid: profile.uid,
	email: profile.email,
	phone: profile.phone,
	workPhone: profile.workPhone,
	firstname: profile.firstname,
	lastname: profile.lastname,
	capabilities: profile.capabilities,
	displayName: profile.displayName,
	name: profile.uid
	```

The above fields will all be available in the `profile` callback.


```

Comment (mike818148) [reply docs/pages/getting-started/providers/sailpoint.mdx:154]: Thanks a lot, this looks perfect, in my case, I also did the same which extend the User Type and also User table schema

File: docs/pages/getting-started/providers/sailpoint.mdx (hunk 6 above)
Comment (mike818148) [reply docs/pages/getting-started/providers/sailpoint.mdx:140]: Thanks a lot for the suggestion, already made the changes, I still keep name: profile.uid. The uid value is like yu-chih.chung, which is more useful as the login account name for SailPoint Identity Secure Cloud. I think display name is less usefuly in this case.

**Review by ndom91 — APPROVED** [review]

**Pull request merged**